src/common/cockpitconf.c

```diff
@@ -21,92 +21,152 @@
+  ssize_t line_len;
+
+  cockpit_conf_loaded = true;
+
+  f = fopen (file_path, "r");
+  if (!f)
     {
-      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        goto out;
-
-      g_message ("couldn't load configuration file: %s: %s", file_path, error->message);
-      ret = FALSE;
-      goto out;
+      if (errno != ENOENT)
+        warnx ("couldn't load configuration file: %s: %m\n", file_path);
+      return false;
     }
 
-  groups = g_key_file_get_groups (key_file, NULL);
-  for (i = 0; groups[i] != NULL; i++)
+  assert (regcomp (&re_section, "^[[:space:]]*\\[ *([^][:space:]]+) *\\][[:space:]]*$", REG_EXTENDED|REG_NEWLINE) == 0);
+  assert (regcomp (&re_keyval, "^[[:space:]]*([^=[:space:]]+)[[:space:]]*=[[:space:]]*(.*)$", REG_EXTENDED|REG_NEWLINE) == 0);
+  assert (regcomp (&re_ignore, "^[[:space:]]*(#.*)?$", REG_EXTENDED|REG_NOSUB) == 0);
+
+  while ((line_len = getline (&line, &line_size, f)) >= 0)
     {
-      gchar **keys;
-      section = g_hash_table_lookup (cockpit_conf, groups[i]);
-      if (section == NULL)
+      regmatch_t matches[3];
+
+      if (regexec (&re_section, line, 3, matches, 0) == 0)
         {
-          section = g_hash_table_new_full (cockpit_str_case_hash, cockpit_str_case_equal, g_free, g_free);
-          g_hash_table_insert (cockpit_conf, g_strdup (groups[i]), section);
+          free (cur_section);
+          cur_section = strlower (strndup (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so));
         }
 
-      keys = g_key_file_get_keys (key_file, groups[i], NULL, &error);
-      g_return_val_if_fail (error == NULL, FALSE);
+      else if (regexec (&re_keyval, line, 3, matches, 0) == 0)
+        {
+          Entry *e;
+
+          if (!cur_section)
+            {
+              warnx ("%s: key=val line not in any section: %s", file_path, line);
+              ret = false;
+              break;
+            }
+
+          e = malloc (sizeof (Entry));
+          e->section = strdup (cur_section);
+          e->key = strlower (strndup (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so));
+          e->value = strndup (line + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
+          e->strv_cache = NULL;
+          /* prepend new Entry to cockpit_conf; that way, later values win over earlier ones in a forward search */
+          e->next = cockpit_conf;
+          cockpit_conf = e;
+        }
 
-      for (j = 0; keys[j] != NULL; j++)
+      else if (regexec (&re_ignore, line, 0, NULL, 0) != 0)
         {
-          gchar *value = g_key_file_get_value (key_file, groups[i], keys[j], &error);
-          g_return_val_if_fail (error == NULL, FALSE);
-          g_hash_table_insert (section, g_strdup (keys[j]), value);
+          warnx ("%s: invalid line: %s", file_path, line);
+          ret = false;
+          break;
         }
-      g_strfreev (keys);
     }
-  g_strfreev (groups);
 
-  g_debug ("Loaded configuration from: %s", file_path);
+  free (line);
```

Comment:
We should check whether `getline` has returned an error.

Reply:
`line` is initialized to `NULL` for that very reason, so that should be fine?

Reply:
`getline` reports error by returning -1 and the loop is stopped in that case. So when the loops ends and `line_len < 0` we need to output a message based on `errno`, no?

Reply:
That's a bit  tricky as `getline` also returns with -1 on EOF. (The example code in the manpage also does this). But we could set `errno = 0` right before, and then check afterwards perhaps.